### PR TITLE
gnuradioPackages.fosphor: init at unstable-2024-03-23

### PR DESCRIPTION
**Pull request by chuangzhu**
## Description of changes

GNU Radio block for RTSA-like spectrum visualization using OpenCL and OpenGL acceleration.

https://projects.osmocom.org/projects/sdr/wiki/Fosphor

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/gnuradio-modules/fosphor/default.nix [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    logLib
+    mpir
+    gmp
+    boost
+    libGL
+    opencl-headers
+    ocl-icd
+    freetype
+    fftwFloat
+  ] ++ lib.optionals enableQt [
```
Comment (doronbehar) [inline pkgs/development/gnuradio-modules/fosphor/default.nix:58]: Maybe we can condition this based on whether `gnuradio.hasFeature "gr-qtgui"`?

File: pkgs/development/gnuradio-modules/fosphor/default.nix [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    repo = "gr-fosphor";
+    rev = "e02a2ea4936324379b02c5a1d4878b2da0961bd9";
+    hash = "sha256-nJEwWBZep5KX8eF+yA8QxDk07pcO7l3OSmKxXAU8KCs=";
+  };
+  disabledForGRafter = "3.11";
+
+  nativeBuildInputs = [
+    cmake
+    pkg-config
+  ] ++ lib.optionals enableQt [
```
Comment (doronbehar) [inline pkgs/development/gnuradio-modules/fosphor/default.nix:44]: Same here.

File: pkgs/development/gnuradio-modules/fosphor/default.nix [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+, python
+, fftwFloat
+, libX11
+, enableQt ? true, qt5
+, enableGLFW ? true, glfw3
+, enablePNG ? true, libpng
+}:
+
+# There is a `gt3.8` branch, but it doesn't build
+# We don't have a disabledForGRbefore, so...
```
Comment (doronbehar) [inline pkgs/development/gnuradio-modules/fosphor/default.nix:24]: That's seems like a nice addition for a prior commit!

**Review by doronbehar — COMMENTED** [review]
Looks great overall, thanks for the contribution!

**Comment by doronbehar** [comment]
Also, it seems CI is failing, due to failed tests in pyqtgraph - not due to this PR. Would be nice to fix those issues before testing this PR.

File: pkgs/development/gnuradio-modules/mkDerivation.nix [hunk 4]
```diff
@@ -7,8 +7,10 @@ mkDerivation:
 args:
 
 # Check if it's supposed to not get built for the current gnuradio version
-if (builtins.hasAttr "disabledForGRafter" args) &&
-(lib.versionAtLeast unwrapped.versionAttr.major args.disabledForGRafter) then
+if ((builtins.hasAttr "disabledForGRafter" args) &&
+(lib.versionAtLeast unwrapped.versionAttr.major args.disabledForGRafter)) ||
+((builtins.hasAttr "disabledForGRbefore" args) &&
+(lib.versionOlder unwrapped.versionAttr.major args.disabledForGRbefore)) then
```
Comment (chuangzhu) [inline pkgs/development/gnuradio-modules/mkDerivation.nix:13]: I wonder if it will be better if we use something like `gnuradioAtLeast` and `gnuradioOlder`, just like `python.pythonAtLeast` `python.pythonOlder`. But if that is the case, we are going to modify all the files under `pkgs/development/gnuradio-modules/`. I would say let's keep that for another PR

Comment (doronbehar) [reply pkgs/development/gnuradio-modules/mkDerivation.nix:13]: > I wonder if it will be better if we use something like `gnuradioAtLeast` and `gnuradioOlder`, just like `python.pythonAtLeast` `python.pythonOlder`. But if that is the case, we are going to modify all the files under `pkgs/development/gnuradio-modules/`. I would say let's keep that for another PR

Sounds great to me! However, we might want to `throw` some message if `disabledForGRafter` is used, because that argument might be used outside of nixpkgs in other people's flakes or alike.

Comment (doronbehar) [reply pkgs/development/gnuradio-modules/mkDerivation.nix:13]: For the record, I'm fine with the way things are now too.

File: pkgs/development/gnuradio-modules/fosphor/default.nix [hunk 5]
```diff
@@ -0,0 +1,83 @@
+{ lib
+, mkDerivation
```
Comment (FliegendeWurst) [inline pkgs/development/gnuradio-modules/fosphor/default.nix:2]: Please format with the new formatter.

**Review by FliegendeWurst — CHANGES_REQUESTED** [review]

**Comment by doronbehar** [comment]
I should have merged this a long time ago.. sorry @chuangzhu for forgetting this. Will merge now when CI is green.
